crates/sui-types/src/gas.rs

```diff
@@ -289,53 +288,45 @@ impl std::fmt::Display for GasCostSummary {
 
 pub fn deduct_gas_legacy(gas_object: &mut Object, deduct_amount: u64, rebate_amount: u64) {
     // The object must be a gas coin as we have checked in transaction handle phase.
-    let gas_coin = GasCoin::try_from(&*gas_object).unwrap();
-    let balance = gas_coin.value();
+    let gas_coin = gas_object.data.try_as_move_mut().unwrap();
+    let balance = gas_coin.get_coin_value_unsafe();
     assert!(balance >= deduct_amount);
-    let new_gas_coin = GasCoin::new(*gas_coin.id(), balance + rebate_amount - deduct_amount);
-    let move_object = gas_object.data.try_as_move_mut().unwrap();
-    // unwrap safe because GasCoin is guaranteed to serialize
-    let new_contents = bcs::to_bytes(&new_gas_coin).unwrap();
-    assert_eq!(move_object.contents().len(), new_contents.len());
-    move_object.update_coin_contents(new_contents);
+    gas_coin.set_coin_value_unsafe(balance + rebate_amount - deduct_amount)
 }
 
 pub fn deduct_gas(gas_object: &mut Object, charge_or_rebate: i64) {
     // The object must be a gas coin as we have checked in transaction handle phase.
-    let gas_coin = GasCoin::try_from(&*gas_object).unwrap();
-    let balance = gas_coin.value();
+    let gas_coin = gas_object.data.try_as_move_mut().unwrap();
```

Comment:
Seems that you could operate on the gas_coin directly

Reply:
I did not understand this comment--is it related to the issue above about exposing `Object::get_coin_value_unsafe`?